Update cctz

alexey-milovidov opened the pull request.

kpozin committed:
[reland] Add basic support for Fuchsia (#202)

- Implement `FuchsiaZoneInfoSource` to read tzif2 files from their designated locations on Fuchsia systems.
- Implement calls to Fuchsia APIs in `local_time_zone()`.
- Disable a time conversion edge case that is unsupported on Fuchsia.

This reverts commit 63b7130d4b3e43e6d09342470fb3732025555f50.

dinord committed:
Remove bazelbuild/rules_cc dependency (#203)

* Remove bazelbuild/rules_cc dependency

The Bazel team recommends using native rules nowadays: https://github.com/bazelbuild/rules_cc/issues/91#issuecomment-756673529
Abseil stopped depending on `rules_cc` in https://github.com/abseil/abseil-cpp/pull/1038/commits/b2387f0b2326efc3fac804426e335a91085839d1.

devbww committed:
Update zoneinfo files to 2021d. (#204)

dinord committed:
Remove a few redundant calls to `unique_ptr.get()`: https://clang.llvm.org/extra/clang-tidy/checks/readability-redundant-smartptr-get.html (#205)

devbww committed:
Update zoneinfo files to 2021e. (#206)

kpozin committed:
[fuchsia] Fix bug in Fuchsia async code (#207)

The existing implementation can interfere with other FIDL code running
on the same thread.

This change ensures that the `async::Loop` does not get attached to any
thread, and therefore does not clear out any thread's dispatcher when
the loop is destroyed.

derekmauro committed:
Fix 32-bit ARM builds (#208)

This patch changes the linking of 32-bit ARM zone info source to use
the same code as the 32-bit X86 zone info source.

This change is requested in
https://github.com/abseil/abseil-cpp/issues/829 by members of the
Microsoft organization who say it fixes the issue.

tachyonwill committed:
Small microptimizations to year_index. (#209)

There are two optimizations. The first is to cache the year index while
stepping through 100 year and 4 year intervals in n_day.

The second is an optimization to the year index calculation itself.
The year_index function had been forcing the modulus result to be
non-negative by adding 400 and taking the modulus again. Even with the
multiply&shift technique that compilers use for integer division with
known divisor, this isn't very cheap. Most of the time it should be
faster to do a conditional-move or branch on the modulus results, though
it is possible there are processors where this isn't faster.

Some microbenchmark results:

name                                                         old time/op             new time/op             delta
BM_Difference_Days                                           14.5ns ± 3%             14.6ns ± 3%     ~             (p=0.387 n=9+9)
BM_Step_Days                                                 10.5ns ± 1%             10.5ns ± 2%     ~             (p=0.645 n=8+8)
BM_GetWeekday                                                1.49ns ± 3%             1.50ns ± 2%     ~             (p=0.520 n=8+8)
BM_NextWeekday                                               13.6ns ± 6%             13.3ns ± 3%     ~             (p=0.050 n=9+9)
BM_PrevWeekday                                               12.6ns ± 3%             12.3ns ± 4%   -2.36%          (p=0.019 n=9+9)
BM_Zone_LoadUTCTimeZoneFirst                                 13.2ns ± 2%             13.0ns ± 4%     ~             (p=0.200 n=8+9)
BM_Zone_LoadUTCTimeZoneLast                                  14.0ns ±13%             13.1ns ± 4%     ~            (p=0.252 n=10+9)
BM_Zone_LoadTimeZoneFirst                                     354µs ± 4%              294µs ± 3%  -16.99%         (p=0.000 n=10+8)
BM_Zone_LoadTimeZoneCached                                   40.2ns ± 7%             39.4ns ± 9%     ~             (p=0.387 n=9+9)
BM_Zone_LoadLocalTimeZoneCached                               136ns ± 6%              136ns ±14%     ~            (p=0.497 n=10+9)
BM_Zone_LoadAllTimeZonesFirst                                 183µs ±29%              117µs ±16%  -36.09%         (p=0.000 n=10+8)
BM_Zone_LoadAllTimeZonesCached                               67.7ns ± 6%             66.2ns ± 4%     ~             (p=0.094 n=9+9)
BM_Zone_TimeZoneEqualityImplicit                             5.82ns ± 4%             5.78ns ± 2%     ~             (p=0.288 n=9+8)
BM_Zone_TimeZoneEqualityExplicit                             2.46ns ± 4%             2.46ns ± 3%     ~            (p=1.000 n=10+9)
BM_Zone_UTCTimeZone                                          2.47ns ± 3%             2.48ns ± 2%     ~             (p=0.666 n=9+9)
BM_Time_ToCivil_CCTZ                                         43.9ns ± 5%             43.5ns ± 3%     ~            (p=0.447 n=10+9)
BM_Time_ToCivil_Libc                                         67.3ns ± 3%             67.4ns ± 1%     ~             (p=0.815 n=9+8)
BM_Time_ToCivilUTC_CCTZ                                       117ns ± 4%               83ns ± 3%  -29.14%         (p=0.000 n=10+9)
BM_Time_ToCivilUTC_Libc                                      40.9ns ± 2%             40.7ns ± 3%     ~             (p=0.888 n=8+9)
BM_Time_FromCivil_CCTZ                                       36.4ns ± 1%             36.2ns ± 2%     ~             (p=1.000 n=8+9)
BM_Time_FromCivil_Libc                                       1.00µs ± 3%             1.00µs ± 4%     ~             (p=0.489 n=9+9)
BM_Time_FromCivilUTC_CCTZ                                    27.2ns ± 0%             26.9ns ± 3%   -1.10%          (p=0.026 n=7+8)
BM_Time_FromCivilDay0_CCTZ                                   61.2ns ± 1%             60.5ns ± 5%     ~             (p=0.370 n=8+9)
BM_Time_FromCivilDay0_Libc                                   1.07µs ± 3%             1.07µs ± 4%     ~            (p=0.780 n=10+9)
BM_Format_FormatTime/0           [%a, %d %b %Y %H:%M:%S %z]   393ns ±14%              367ns ± 3%     ~            (p=0.079 n=10+9)
BM_Format_FormatTime/1           [%d %b %Y %H:%M:%S %z    ]   312ns ± 3%              314ns ± 4%     ~            (p=0.515 n=10+8)
BM_Format_FormatTime/2           [%Y-%m-%d%ET%H:%M:%E*S%Ez]   307ns ± 3%              306ns ± 3%     ~             (p=0.730 n=9+9)
BM_Format_FormatTime/3           [%Y-%m-%d%ET%H:%M:%S%Ez  ]   279ns ± 1%              277ns ± 4%     ~             (p=0.536 n=7+9)
BM_Format_FormatTime/4           [%Y-%m-%d%ET%H:%M:%S     ]   236ns ± 4%              238ns ± 3%     ~             (p=0.489 n=9+9)
BM_Format_FormatTime/5           [%Y-%m-%d                ]   126ns ± 3%              126ns ± 3%     ~             (p=0.963 n=9+8)
BM_Format_ParseTime/0            [%a, %d %b %Y %H:%M:%S %z]  1.41µs ± 3%             1.40µs ± 4%     ~             (p=0.546 n=9+9)
BM_Format_ParseTime/1            [%d %b %Y %H:%M:%S %z    ]  1.15µs ± 9%             1.17µs ±16%     ~            (p=0.720 n=9+10)
BM_Format_ParseTime/2            [%Y-%m-%d%ET%H:%M:%E*S%Ez]   305ns ± 4%              302ns ± 3%     ~            (p=0.447 n=10+9)
BM_Format_ParseTime/3            [%Y-%m-%d%ET%H:%M:%S%Ez  ]   272ns ± 1%              271ns ± 2%     ~             (p=0.743 n=9+8)
BM_Format_ParseTime/4            [%Y-%m-%d%ET%H:%M:%S     ]   165ns ± 2%              173ns ±16%     ~            (p=0.780 n=9+10)
BM_Format_ParseTime/5            [%Y-%m-%d                ]   113ns ± 4%              114ns ± 4%     ~             (p=0.606 n=9+8)

devbww committed:
Add missing #include <ios> for 'std::left'. (#215)

derekmauro committed:
Replace NULL with nullptr (#216)

This change is in support of the go/NULL-termination LSC

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ClickHouse/cctz?pullRequest=3) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/ClickHouse/cctz?pullRequest=3) before we can accept your contribution.<br/>**0** out of **5** committers have signed the CLA.<br/><br/>:x: kpozin<br/>:x: derekmauro<br/>:x: devbww<br/>:x: dinord<br/>:x: tachyonwill<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ClickHouse/cctz?pullRequest=3) it.</sub>

devbww committed:
Update zoneinfo files to 2022a. (#217)

devbww committed:
Update the script for generating testdata/zoneinfo. (#219)

Provide a switch to use github.com/JodaOrg/global-tz (which reinstates
rules removed from recent versions of the IANA database) as the tzdata
source.

devbww committed:
Add a NextTransition.Scan test case to time_zone_lookup_test. (#218)

Check the shape of every transition in every timezone.  This gives
extra confidence that we have interpreted the zoneinfo data correctly.
It also provides a convenient location for temporary edits to do
things like examine all the UTC offsets or zone abbreviations in use.

devbww committed:
Limit testing of extended format/parse specifiers to GLIBC (#220)

Restrict testing of some extended format/parse specifiers (e.g.,
%k, %l, and %P) to platforms that define `__GLIBC__`.  This means
the tests can run with alternative Linux C run-time libraries,
like https://musl.libc.org/, that do not support those extensions.

devbww committed:
Limit MakeTime.LocalTimeLibC test to GLIBC (#221)

Further restrict the environments where we know what to expect from
`mktime()`; in particular to `__GLIBC__`.  This means the test will
also be skipped when run with alternative Linux C run-time libraries,
like https://musl.libc.org/.

rschu1ze commented:
Not sure but I guess we should not reuse the same PR for multiple merges from upstream.

I have created a dedicated PR at https://github.com/ClickHouse/cctz/pull/4. Let's close this one.
